andela/ah-jarvis
# #159100088 Notifications

#### What does this PR do?
Sends email notifications to a user's followers when a new article is created. Enables a user to receive notifications when a favourited article has a new comment.

#### Description of Task to be completed?
- Write tests for the feature
- Create email sending pipeline that includes celery, Rabbitmq and a flower dashboard
- Allow user followers to receive an email when a new article is posted
- Allow users to receive notification on favourited articles when a new comment is posted
- Allow users to opt in and out of the notifications feature

##### Endpoints

METHOD | ENDPOINT | FUNCTIONALITY 
------------ | -------------|--------------
GET | /api/notifications/ | Gets all the notifications for logged in user
PUT | /api/user/  | Subscribe/unsubscribe to notifications``` { user : { get_notifications:true } } ``` 

#### How should this be manually tested?
- Clone this repo

```
cd ah-jarvis

git fetch

git checkout ft-notifications-159100088
  
create and activate a virtual env 

pip install -r requirements.txt

brew install rabbitmq

python manage.py migrate

coverage run manage.py test

```
- When all the tests pass, it means that the feature works. 
- Run celery on a different terminal
```
celery worker -A authors -l info
```

##### Test using postman
- Launch Postman App
- Create three users `user1`, `user2`, `user3`
- Make `user2` follow `user1`
- Make `user1` create an article
- Check email for `user2`
- Make `user3`favourite article by `user1`
- Make `user2` comment on article by `user1`
- Check email for `user3`

#### Screenshots
- Subscribe
<img width="1185" alt="screen shot 2018-08-16 at 19 55 25" src="https://user-images.githubusercontent.com/13402544/44222994-002acb80-a18f-11e8-9eb3-b57f7004ba87.png">
- Get Notifications
<img width="1181" alt="screen shot 2018-08-17 at 08 02 59" src="https://user-images.githubusercontent.com/38427824/44248654-8e856880-a1f4-11e8-9736-e5160b0ec625.png">


#### Any background context you want to provide?
None

#### What are the relevant pivotal tracker stories?
[#159100088](https://www.pivotaltracker.com/story/show/159100088/comments/193133209) Users should be able to receive notifications
